### PR TITLE
address issue #706

### PR DESCRIPTION
**Pull request by mabi**
some of the changes were already done by @glynhudson but should not cause any trouble

### CODE REVIEW

File: Modules/feed/engine/CassandraEngine.php [hunk 1]
```diff
@@ -55,8 +56,11 @@ public function create($feedid,$options)
     */
     public function delete($feedid)
     {
-    		$feedid = (int) $feedid;
-        $this->execCQL("DROP TABLE feed_".$feedid);
+    	$feedid = (int) $feedid;
+        $feedname = $this->feedtableToDrop($feedid);
+        if(feedname){
```
Comment (mabi) [inline Modules/feed/engine/CassandraEngine.php]: @TrystanLea  sure! sorry

**Comment by mabi** [comment]
Hello @TrystanLea did you check my fix? Thanks

**Comment by Paul-Reed** [comment]
Where are we this pls?
So far @TrystanLea  merged the initial commit for Cassandra, but it has not progressed further.
We appear to have Cassandra offered as an option in emoncms master, yet it's not functional.
It would be nice to either develop or ditch!

Paul

**Comment by mabi** [comment]
@Paul-Reed it would be nice that this PR is included in the main branch if the modifications I made to the docs and code are OK for you. The code is in production since I opened the PR and I think that with this fixes Cassandra support is fully usable.

**Comment by Paul-Reed** [comment]
Hi Marco, I'm leaving this to Trystan to progress, because I'm not in a position to test it myself, and also aware that Trystan made the initial cassandra merge decision.

I commented above as a reminder, as I'm aware that your revised PR has been there a while now, and I didn't want it to get overlooked!

**Comment by mabi** [comment]
@TrystanLea sorry to bother you, any progress on this? It would be great if you include this on the main branch to complete the first Cassandra PR, code is in production with no issues since then. Thank you

**Comment by Paul-Reed** [comment]
@mabi The big problem I see, is that there doesn't appear to be a version of Cassandra which is specifically compiled for armhf - and as the vast majority of emoncms users use raspberry pi's, it makes installing it a complicated process.
Also in issue #706 I suggested that the Cassandra read.me would need to be expanded to cover the FULL 'cassandra' installation process, and not just a few development notes.

IMO this integration is not the best option for emoncms for the reasons above, however @TrystanLea and others may have a different view.

Paul

**Comment by pb66** [comment]
Can I ask a question about Casandra in emoncms?

What is the "cassandra namespace"?

Is this simply a folder for cassandra data?

Could we avoid using the toplevel name "emoncms" this doesn't indicate it's used for cassandra and it blocks /var/lib/emoncms being used to hold all emoncms sub-folders phpfina and phptimeseries etc. Could we please use "cassandra" for the cassandra namespace? 

ideally we would use 
/var/lib/emoncms/cassandra
/var/lib/emoncms/phpfina
/var/lib/emoncms/phptimeseries

so all emoncms stuff is grouped and possibly we could think about reducing the 4 locations to a single /var/lib/emoncms with fixed sub-directories.

**Comment by mabi** [comment]
@Paul-Reed Cassandra integration is optional and already in the master branch, this PR is just a bug fix and improvement over the old PR, IMHO pulling this will close the issue and leave Cassandra support in a usable state. Right now is half finished.

**Comment by mabi** [comment]
@pb66 for more info have a look here [https://github.com/mabi/emoncms/blob/dev-cassandra/docs/Cassandra.md](Cassandra.md), Cassandra does not need a folder under /var/lib/emoncms

**Comment by pb66** [comment]
@mabi [that link](https://github.com/emoncms/emoncms/compare/Cassandra.md?expand=1) doesn't work. I assume you mean https://github.com/mabi/emoncms/blob/dev-cassandra/docs/Cassandra.md which doesn't offer much to answer my questions.

What is the "cassandra namespace"?

Is this simply a folder for cassandra data?

If "Cassandra does not need a folder under /var/lib/emoncms" what is [this](
https://github.com/emoncms/emoncms/blob/master/default.settings.php#L66-L68)
```
'cassandra'=>array(
            'keyspace' => 'emoncms'
        )
```
what does this relate to? is it in-memory? is there a cassandra folder as per mysql? Where is the data persisted? Where the setting has been added under the feed engine storage paths, it looks like a partial path, perhaps it needs clarification. Perhaps a simple comment block between the paths and the namespace might help considerably, but I'm guessing.

If it is not clear to me, I am unlikely to be the last person to ask. Reviewing your proposed changes requires an understanding of it. I have not used cassandra so I am asking you for some simple clarification.

**Comment by mabi** [comment]
@pb66 sorry for the broken link, I mixed description and URL in md, if you copy paste it it should work. It points to the updated version in my PR, that is the reason I wolud like to have it merged, so that Cassandra support is slightly more documented and implemented.
[https://github.com/mabi/emoncms/blob/dev-cassandra/docs/Cassandra.md](https://github.com/mabi/emoncms/blob/dev-cassandra/docs/Cassandra.md)
Cassandra is a distributed database, so where the files are stored is not an issue. A keyspace is a database name in Cassandra parlance. Don't hesitate to ask, it is a busy moment so I pointed you to the link hoping it wolud clarify.

**Comment by pb66** [comment]
"so where the files are stored is not an issue." 

I would argue it is whole point...

Is there anything persisted to the local install? if so where? I have had a brief browse and google searches of cassandra and learnt it is a masterless cluster of servers and the data storage is shared amongst them. Is you proposal a cluster of one? If not who are we clustering with? who is my data being shared with? who else's data is going to be stored on my server?

I see a ref to /var/lib/cassandra/data in the config yaml file. Is this where the data is stored?

@Paul-Reed's point about raspberry pi users is very valid, 9 out of 10 users are Pi based, How will this perform with an RO OS?

IMO the first part of this was prematurely merged, I think a fully working "cassandra" feature branch should have been set up for this. But before I can think about how, where and when I need to understand why!

What are the benefits of using cassandra? Are you suggesting one big network? Is there a large cassandra userbase that might want emoncms? I do not see the likes of apple, facebook and ebay wanting emoncms and they seem to be the key users.

If there is good reason to include it and it's done in a thoughtful way, I'm all in, but if this a very niche need I see it as unnecessary complexity and a better route would be for Marco to maintain a seperate "cassandra" repo, don't forget it's not just about getting it in to emoncms, it's about ongoing maintenance and not breaking the cassandra implementation in future development to emoncms. If none of the emoncms devs are familiar with cassandra or have a desire to ensure the cassandra implementation is maintained, it will rely on Marco to keep coming back and fixing it or it becomes dead code.

In the settings.php I see only the enable/disable cassandra engine and the name space. I see no mention of hosts, ports, authentication, urls etc etc, just on or off. I know most/all of this will be managed in Cassandra itself, but some insight to what emoncms is signing up for when that cassandra engine is enabled is required.

I need to know more to get on-board. I'm not against cassandra being included, but the biggest stumbling block for me right now is a lack of info. Right now I am not interested enough to spend loads of time researching it, but I might if I'm sold the idea....

**Comment by mabi** [comment]
@pb66 I think that this discussion is out of context: Cassandra support is only for server side installation of emoncms, not on emonpi obviously. You connect to an external DB not on your local storage.

**Comment by mabi** [comment]
@pb66 regarding the feature branch part of your comment I agree with you, it should be a feature branch, but right now we are in the middle, either remove Cassandra from main or complete with missing code. In my own installation Cassandra is used to store a lot of data with no single point of failure.

**Comment by pb66** [comment]
"Cassandra support is only for server side installation of emoncms, not on emonpi obviously."
emonpi is a emoncms server and I'm finding nothing about this "obvious".

"You connect to an external DB not on your local storage."
so where is that? surely the emoncms server cannot just consume network storage without providing any in return?

"either remove Cassandra from main or complete with missing code."
Currently removing it gets my vote, but I do not want to dampen any enthusiasm, I just don't think it's justified or ready yet. 

I really am trying to understand so I can  support including Cassandra.

"I think that this discussion is out of context:"

In which case go on the forum and start a discussion or include a bit more info in the emoncms/docs on using cassandra, it's benefits to emoncms and how it works. "Cassandra support" possibly means nothing to good chunk of the community, me included. Selling the idea will get you support and pave the way to it's inclusion.

**Comment by mabi** [comment]
@pb66 I didn't want to upset anyone, I just wanted to help, I started a discussion on the forum [https://community.openenergymonitor.org/t/cassandra-support-in-emoncms/5725](https://community.openenergymonitor.org/t/cassandra-support-in-emoncms/5725)

**Comment by pb66** [comment]
"I didn't want to upset anyone, I just wanted to help,"

You haven't upset anyone, likewise I hope you haven't taken offence to my questions. I would just like to understand what the advantages and the consequence of Cassandra are for emoncms to make an informed decision.

**Comment by TrystanLea** [comment]
I've merged this in, given that the original was a half way solution. I've also removed the requirement to update the settings file on live emonpi installations. 

Thinking about this a bit more I think we should remove this from the core but document its availability for anyone interested. Would you be able to maintain a repository on your github @mabi with the CassandraEngine.php file and a readme documenting how to use it with emoncms. i.e drop in the emoncms/Modules/feed/engine folder, change settings.php with required changes?

**Comment by TrystanLea** [comment]
There's a bigger question here about how we support modularity for feed engines as well as what to do about old feed engines like Histogram.php and PHPFiwa.php that are now depreciated but kept in for backwards compatibility..

**Comment by mabi** [comment]
@TrystanLea thanks for addressing this. I think that your proposal is good, just reserve the engine code for cassandra withiut including the actual engine and I will maintain a repository with the engine as a 'drop in', like I was doing before this merge: I was checking out the official version and replacing cassandra engine with my version.

**Comment by TrystanLea** [comment]
Ok great, we could add a .gitignore in the feed module so that copying in the engine file does not show up on the emoncms git.  What exactly do you mean by "just reserve the engine code for cassandra"?

**Comment by mabi** [comment]
@TrystanLea I mean keeping const `CASSANDRA = 10;` in Lib/enum.php

**Comment by TrystanLea** [comment]
Ah cool, yes fine, lets do that.

**Comment by TrystanLea** [comment]
Let me know when you have the repository and readme up and running and Il make the change to the core as well as adding a node in the readme as to the existence of the Cassandra option

**Comment by mabi** [comment]
@TrystanLea the repository is up at https://github.com/mabi/emoncms-cassandra 
I also aligned mabi/emoncms (master and stable) to emoncms/emoncms and I plan to maintain a cassandra branch merging periodically from emoncms master

**Pull request merged**